Many changes

gear4s opened the pull request:
NOTE: This PR requires the latest version of discord.py

Major changes:
- Discord.py update:
  - [#](commits/b9e5adb23f74b0ff07eb4e6adbec0f78db3ddee1) Discord.py rewrite released a [breaking change](https://gist.github.com/Rapptz/288294ca99fa1f042f0e39a92ddd88eb#news-post) which changed functionality of the help command, and the loading and unloading of cogs.
- Cogs:
  - [#](a4e516cb60167dc4e3ca90c454a72990516a1eda) Base cog class now has default setup method
  - [#](dcf236af3262d5d12eb2414402f6fd97cd937893) load_extension now checks if there is an attribute named Cog in a cog file, and loads setup from that (or the file if the Cog.setup attribute doesnt exist)
  - [#](f0efdb688aecd7083b1bb5a192a0974bc44e5a2f) unload_extension now checks if there is an attribute named Cog in a cog file, and runs teardown from that (or the file if the Cog.teardown attribute doesnt exist)
  - [#](b9e5adb23f74b0ff07eb4e6adbec0f78db3ddee1) Use a new cog, "jishaku", from the pip repository, which allows for ease of development of the bot. In future, use this cog for public commands (such as source code for commands)
- Core:
  - [#](8b14ed025239a221e1efdd27208de6c903eba67d) Config is now a metaclass instead of an inherited class
- Database:
  - [#](69ec004d38194e8934e27f987ea915ec62b52b9b) Build query when .build is used, instead of building as functions are executed

Minor:
- [#](45454cbaf1e0e286e684d4574cd7676df9ee9892) Abstracted the help message into base cog class
- [#](db8b60e12b24b040479f8bc09d6ac16438d1786e) Trigger configuration setup
- [#](2a0720eec98c82c653c9332ea78ffe61fe9a6d2e) Mention user before the reply content in ctx.reply()
- [#](f3aebbfb73d637232230b636903bd50c469c4edf) Renamed stats.py to db_events.py
- [#](c8c51b10a72875f32e3e10063ea9d83e3d921c68) Exposed delete from database library. Also added a .run() method that takes a cursor as an argument.
- [#](0f240f4e009fe15d00b92a92219bb0812ddd84b7) Add ability to use composite primary keys in database
- [#](1586f8bef3c277d94f9274927a93e993f4af1a11) `!n.say` now uses a #channel mention and uses the `message` kwarg

Bugfixes:
- [#](c7558bec9be293bd4b94dbd1db34503b1dd54174) `!n.cogs list` no longer raises
- [#](1586f8bef3c277d94f9274927a93e993f4af1a11) `!n.status` now works
- [#](c7558bec9be293bd4b94dbd1db34503b1dd54174) Fix `!n.cogs load` and `!n.cogs unload`
- [#](83574e7d68d2bd7e4d65a3719b669186c24e3b84) `!n.help` should no longer raise about NoneType and .lower()

gear4s committed:
Update tags.py

gear4s committed:
Add a default setup method for cogs

gear4s committed:
Change method of cog loading by using a class named Cog and inserting setup as a class function
Update all Cogs to reflect this change

gear4s committed:
Update kwargs instead of whatever the heck used to be there

gear4s committed:
Abstract help sending if no subcommands passed

gear4s committed:
Singleton as metaclass insetad of inherited class

gear4s committed:
Remove cogs import from system modules for a full reload and unload

gear4s committed:
Revert "Remove cogs import from system modules for a full reload and unload"

This reverts commit f1af37d

gear4s committed:
Start adding configuration for triggers

gear4s committed:
Fix cog listing

gear4s committed:
Fix say and nowplaying command, and renamed nowplaying to status (reflects API)

gear4s committed:
Fix cog loading and unloading

gear4s committed:
Customize the method for unloading cogs so that it can use the teardown method in the cog itself

gear4s committed:
Add a TODO and add config cog

gear4s committed:
Mention user before content in reply, and remove mention in CheckFailure error

gear4s committed:
Move stats.py into db_events.py

gear4s committed:
Add a .run() method to BaseQuery database class

gear4s committed:
Use new .run in db_events.py

gear4s committed:
Expose a delete method from db_util

gear4s committed:
Actually build the query in the build property instead of building them in the setting functions

gear4s committed:
Ability to check composite primary keys in where clause

gear4s committed:
Fix index error and quote the keys to check

gear4s committed:
Add ability to insert raw sql into inserts

gear4s committed:
Remove append_sql and add or_update for duplicate key updating

gear4s committed:
Fix bug with config leading to error in help command

safaci2000 commented on database/methods/delete.py:
```diff
@@ -0,0 +1,23 @@
+
+
+class Delete(BaseQuery):
+    def __init__(self, table_name, where=None):
```
isn't this dangerous, what happens if invoked and where condition is None?

basically want to avoid delete from table x; 

delete operations becomes a bit more dangerous, so have to be more careful.

gear4s replied:
then it would delete everything from the table x. seeing as no one except the devs should have access to this, i dont see a problem with it, i'm trying to keep the wrapper as close to the sql language as possible

gear4s committed:
Breaking changes from Discord.py:
- Changed HelpFormatter to HelpCommand
- Changed from overriding load_extension and unload_extension to overriding _load_from_module_spec and _call_module_finalizers
- Changed handling errors with cog loading

Other changes:
- Use a cog named "jishaku" when testing bot. NOTE: Requires discord.py installed to use
